File: leetcode/go/682.baseball-game.go
func calPoints(operations []string) int {
    s := make(stack,0)
    for _, o := range operations {
        switch o {
        case "+":
            l := len(s)
            s = s.Push(s[l-1] + s[l-2])
            break
        case "D":
            s = s.Push(2 * s[len(s)-1])
            break
        case "C":
            s, _ = s.Pop()
            break
        default:
            x, _ := strconv.Atoi(o)
            s = s.Push(x)
        }
    }
    ans := 0
    for _, score := range s {
        ans += score
    }
    return ans
}

type stack []int

func (s stack) Peek(v int) int {
    return s[len(s)-1]
}

func (s stack) Push(v int) stack {
    return append(s, v)
}

func (s stack) Pop() (stack, int) {
    l := len(s)
    return  s[:l-1], s[l-1]
}

File: leetcode/rust-leetcode/src/682.baseball-game.rs
impl Solution {
    pub fn cal_points(operations: Vec<String>) -> i32 {
        let mut stack = vec![];
        let mut score = 0;
        for o in operations.iter() {
            match o.as_str() {
                "+" => {
                    let l = stack.len();
                    score = stack[l-1] + stack[l-2];
                    stack.push(score);
                },
                "C" => {stack.pop();},
                "D" => {stack.push((*stack.last().unwrap()) << 1);}
                x @ _ => {
                    stack.push((x.parse::<i32>().unwrap()))
                }
            }
        }
        stack.iter().sum()
    }
}

